crates/shared/src/price_estimation/competition.rs

```diff
@@ -0,0 +1,148 @@
+use super::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use anyhow::{anyhow, Result};
+use futures::future;
+
+/// Price estimator that pulls estimates from various sources
+/// and competes on the best price.
+pub struct CompetitionPriceEstimator {
+    inner: Vec<(String, Box<dyn PriceEstimating>)>,
+}
+
+impl CompetitionPriceEstimator {
+    pub fn new(inner: Vec<(String, Box<dyn PriceEstimating>)>) -> Self {
+        assert!(!inner.is_empty());
+        Self { inner }
+    }
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for CompetitionPriceEstimator {
+    async fn estimates(&self, queries: &[Query]) -> Vec<Result<Estimate, PriceEstimationError>> {
+        let all_estimates =
+            future::join_all(self.inner.iter().map(|(name, estimator)| async move {
+                (name, estimator.estimates(queries).await)
+            }))
+            .await;
+
+        queries
+            .iter()
+            .enumerate()
+            .map(|(i, query)| {
+                all_estimates
+                    .iter()
+                    .filter_map(|(name, estimates)| match &estimates[i] {
+                        Ok(estimate) => Some((name, estimate)),
+                        Err(err) => {
+                            tracing::warn!(
+                                estimator_name = %name, ?query, ?err,
+                                "price estimation error",
+                            );
+                            None
+                        }
+                    })
+                    .filter_map(|(name, estimate)| {
+                        match estimate.price_in_sell_token_rational(query) {
+                            Some(price) => Some((name, estimate, price)),
+                            None => {
+                                tracing::warn!(
+                                    estimator_name = %name, ?query, ?estimate,
+                                    "price estimate with zero amounts",
+                                );
+                                None
+                            }
+                        }
+                    })
+                    .max_by_key(|(_, _, price)| price.clone())
+                    .ok_or_else(|| {
+                        PriceEstimationError::Other(anyhow!("no successful price estimates"))
+                    })
```

Comment:
Probably fine more now but this loses information in some sense by making ever PriceEstimationError into Other. For single price estimators we differentiate for example no liquidity from Other which was supposed to be something like internal error. But here even if all estimators return NoLiquidity it will return Other. This might cause the error to be error logged when it shouldn't or to give the frontend worse information on why an error occurred. Not sure how important all of these points are.

Reply:
Hmm, good point. Choosing the "correct" error out of `n` errors is a tough problem. What we can do is try and find the "common" error across price estimators. For example if we have:
- `InsufficientLiquidity + InsufficientLiquidty + Other` = `InsufficientLiquidity`
- `InsufficientLiquidyt + ZeroAmount + UnsupportedToken` = `Other`

Reply:
Addressed with: https://github.com/gnosis/gp-v2-services/pull/1388